stl/inc/ranges

```diff
@@ -728,6 +738,433 @@ namespace ranges {
+            }
+        }
+
+        // clang-format off
+        constexpr auto begin() const // Constraints per P/R of LWG issue unnumbered as of 2020-08-31
+            requires random_access_range<const _Vw> && sized_range<const _Vw> {
+            // clang-format on
+            const auto _Offset = (_STD min)(_RANGES distance(_Range), _Count);
+            return _RANGES begin(_Range) + _Offset;
+        }
+
+        // clang-format off
+        constexpr auto end() requires (!_Simple_view<_Vw>) {
+            // clang-format on
+            return _RANGES end(_Range);
+        }
+
+        constexpr auto end() const requires range<const _Vw> {
+            return _RANGES end(_Range);
+        }
+
+        constexpr auto size() requires sized_range<_Vw> {
+            const auto _Size          = _RANGES size(_Range);
+            const auto _Count_as_size = static_cast<range_size_t<_Vw>>(_Count);
+            if (_Size < _Count_as_size) {
+                return range_size_t<_Vw>{0};
+            } else {
+                return static_cast<range_size_t<_Vw>>(_Size - _Count_as_size);
+            }
+        }
+
+        constexpr auto size() const requires sized_range<const _Vw> {
+            const auto _Size          = _RANGES size(_Range);
+            const auto _Count_as_size = static_cast<range_size_t<_Vw>>(_Count);
+            if (_Size < _Count_as_size) {
+                return range_size_t<_Vw>{0};
+            } else {
+                return static_cast<range_size_t<_Vw>>(_Size - _Count_as_size);
+            }
+        }
+    };
+
+    template <class _Rng>
+    drop_view(_Rng&&, range_difference_t<_Rng>) -> drop_view<views::all_t<_Rng>>;
+
+    namespace views {
+        // VARIABLE views::drop
+        class _Drop_fn {
+        private:
+            enum class _St { _Empty, _Preserve, _Drop_view };
+
+            template <class _Rng>
+            _NODISCARD static _CONSTEVAL _Choice_t<_St> _Choose() noexcept {
+                using _Ty = remove_cvref_t<_Rng>;
+
+                if constexpr (_Is_specialization_v<_Ty, empty_view>) {
+                    return {_St::_Empty, true};
+                } else if constexpr (_Reconstructible_range<_Rng>) {
+                    return {_St::_Preserve,
+                        noexcept(_Ty{_RANGES begin(_STD declval<_Rng&>()) + _RANGES distance(_STD declval<_Rng&>()),
+                            _RANGES end(_STD declval<_Rng&>())})};
+                } else {
+                    return {_St::_Drop_view, noexcept(drop_view{_STD declval<_Rng>(), range_difference_t<_Rng>{0}})};
+                }
+            }
+
+            template <class _Rng>
+            static constexpr _Choice_t<_St> _Choice = _Choose<_Rng>();
+
+            template <_Integer_like _Ty>
+            class _Partial : public _Pipe::_Base<_Partial<_Ty>> {
+            private:
+                _Ty _Length;
+
+                constexpr explicit _Partial(_Ty _Length_) noexcept : _Length{_Length_} {}

```

Comment:
General question: would being an aggregate improve codegen in this pattern?

Reply:
It's hard to devise use cases where these partially applied range adaptors don't get optimized out of existence - they should appear _very_ rarely on binary interfaces. Then again, they're largely unspecified function objects as far as users are concerned so we can get away with nearly anything. (You may recall from LWG review that the specification of the composition machinery is largely handwavy to encourage implementors to experiment with different approaches.)

I'll make these aggregates for now - the only downside is needing to explicitly initialize the base class with `{}` - and we can change them later if a reason to do so appears.

Reply:
> I'll make these aggregates for now - the only downside is needing to explicitly initialize the base class with `{}` - and we can change them later if a reason to do so appears.

Good news: I can use designated initializers to avoid explicitly initializing the base class, so this is a very nice overall simplification.

Bad news :Aggregates with base classes are apparently not ABI-aggregate, the compiler still returns these by hidden-pointer-argument.

I think I'll keep them aggregates nonetheless for simplicity.

Reply:
(I made a similar change to the `filter_view` PR as well.)